lib/sass/script/functions.rb

```diff
@@ -2078,17 +2092,50 @@ def map_get(map, key)
     # @example
     #   map-merge(("foo": 1), ("bar": 2)) => ("foo": 1, "bar": 2)
     #   map-merge(("foo": 1, "bar": 2), ("bar": 3)) => ("foo": 1, "bar": 3)
+    #   map-merge(("foo": 1, "bar": 2), bar, 3) => ("foo": 1, "bar": 3)
+    #   map-merge((a: (b: c)), a, b, d) => (a: (b: d))
     # @overload map_merge($map1, $map2)
     #   @param $map1 [Sass::Script::Value::Map]
     #   @param $map2 [Sass::Script::Value::Map]
     # @return [Sass::Script::Value::Map]
     # @raise [ArgumentError] if either parameter is not a map
-    def map_merge(map1, map2)
+    # @overload map_merge($map1, $keys..., $map2)
+    #   @param $map1 [Sass::Script::Value::Map]
+    #   @param $keys [[Sass::Script::Value]]
+    #   @param $map2 [Sass::Script::Value::Map]
+    # @return [Sass::Script::Value::Map]
+    # @raise [ArgumentError] if either parameter is not a map
+    def map_merge(map1, *keys)
       assert_type map1, :Map, :map1
+      map2 = keys.pop
+      if map1.is_a? Sass::Script::Value::List
+        map1 = map(map1.to_h)
+      end
       assert_type map2, :Map, :map2
-      map(map1.to_h.merge(map2.to_h))
+      map1.recursive_merge(keys, map2)
     end
     declare :map_merge, [:map1, :map2]
+    declare :map_merge, [:map1], :var_args => true
+
+    # @example
+    #   map-set((a: 1), b, 2) => (a: 1, b: 2)
+    #   map-set((a: (a: 1)), a, b, 2) => (a: (a: 1, b: 2))
+    #   map-set((), a, b, c) => (a: (b: c))
+    # @overload map-set($map, $keys..., $value)
+    #   @param $map1 [Sass::Script::Value::Map]
+    #   @param $keys [[Sass::Script::Value::Base]]
+    #   @param $value [Sass::Script::Value::Base]
+    # @return [Sass::Script::Value::Map]
+    # @raise [ArgumentError] if the first parameter is not a map
+    def map_set(map, *keys)
+      assert_type map, :Map, :map
+      new_value = keys.pop
+      if map.is_a? Sass::Script::Value::List
+        map = map(map.to_h)
+      end
+      map.recursive_set(keys, new_value)
+    end
+    declare :map_set, [:map, :keys, :value], :var_args => true
```

Comment:
This is the wrong `declare` format. This should have the same structure as `map-merge` and `map-get`.